File: frontend/src/services/apiClient.ts
/**
 * HTTP client abstraction for communicating with PharmaSecure Web API.
 * Automatically injects X-Correlation-ID and Authorization headers.
 */
export async function apiClient<T>(endpoint: string, options: RequestInit = {}): Promise<T> {
  const token = localStorage.getItem('auth_token');

  // Generate or forward Correlation ID for end-to-end tracing (Rule 09, Rule 08)
  const correlationId = crypto.randomUUID ? crypto.randomUUID() : Math.random().toString(36).substring(2);

  const headers: HeadersInit = {
    'Content-Type': 'application/json',
    'X-Correlation-ID': correlationId,
    ...(token ? { Authorization: `Bearer ${token}` } : {}),
    ...options.headers,
  };

  const response = await fetch(endpoint, {
    ...options,
    headers,
  });

  if (!response.ok) {
    const errorData = await response.json().catch(() => ({
      status: response.status,
      title: response.statusText,
    }));
    throw errorData;
  }

  return response.json() as Promise<T>;
}

File: mobile/lib/core/network/api_client.dart
import 'dart:convert';
import 'package:http/http.dart' as http;

/// API Client for mobile application communicating with PharmaSecure Web API.
/// Mobile MUST NOT connect directly to SQL Server (Rule 01, Section 3).
class ApiClient {
  final String baseUrl;
  final http.Client _httpClient;

  ApiClient({required this.baseUrl, http.Client? client})
      : _httpClient = client ?? http.Client();

  Future<Map<String, dynamic>> get(
    String endpoint, {
    String? token,
    String? correlationId,
  }) async {
    final uri = Uri.parse('$baseUrl$endpoint');
    final headers = {
      'Content-Type': 'application/json',
      'X-Correlation-ID': correlationId ?? DateTime.now().millisecondsSinceEpoch.toString(),
      if (token != null) 'Authorization': 'Bearer $token',
    };

    final response = await _httpClient.get(uri, headers: headers);

    if (response.statusCode >= 200 && response.statusCode < 300) {
      return jsonDecode(response.body) as Map<String, dynamic>;
    } else {
      throw Exception('API Error: ${response.statusCode} - ${response.body}');
    }
  }

  void dispose() {
    _httpClient.close();
  }
}
